core/src/main/java/org/apache/calcite/rel/rules/IntersectToSemiJoinRule.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that translates a {@link org.apache.calcite.rel.core.Intersect}
+ * to a series of {@link org.apache.calcite.rel.core.Join} that type is
+ * {@link org.apache.calcite.rel.core.JoinRelType#SEMI}.
+ */
+@Value.Enclosing
+public class IntersectToSemiJoinRule
+    extends RelRule<IntersectToSemiJoinRule.Config>
+    implements TransformationRule {
+
+  /** Creates an IntersectToSemiJoinRule. */
+  protected IntersectToSemiJoinRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Intersect intersect = call.rel(0);
+    if (intersect.all) {
+      return; // nothing we can do
+    }
+
+    final RelBuilder builder = call.builder();
+    final RexBuilder rexBuilder = builder.getRexBuilder();
+    final RelDataType oriRowType = intersect.getRowType();
+
+    List<RelNode> inputs = intersect.getInputs();
+    if (inputs.size() != 2) {
+      return;
+    }
+
+    RelNode left = inputs.get(0);
+    RelNode right = inputs.get(1);
+
+    List<RexNode> conditions = new ArrayList<>();
+    int fieldCount = left.getRowType().getFieldCount();
+
+    // Converting the join condition to
+    // AND(=(COALESCE($0, 0), COALESCE($1, 0)), =(IS NULL($0), IS NULL($0))
+    // allows us to use HashJoin
+    for (int i = 0; i < fieldCount; i++) {
+      conditions.add(
+          rexBuilder.makeCall(
+          SqlStdOperatorTable.EQUALS,
+          rexBuilder.makeCall(SqlStdOperatorTable.COALESCE,
+              rexBuilder.makeInputRef(
+                  left.getRowType().getFieldList().get(i).getType(), i),
+              builder.literal(0)),
+          rexBuilder.makeCall(SqlStdOperatorTable.COALESCE,
+              rexBuilder.makeInputRef(
+                  right.getRowType().getFieldList().get(i).getType(), i + fieldCount),
+              builder.literal(0))));
+
+      conditions.add(
+          rexBuilder.makeCall(
+          SqlStdOperatorTable.EQUALS,
+          rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL,
+              rexBuilder.makeInputRef(
+                  left.getRowType().getFieldList().get(i).getType(), i)),
+          rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL,
+              rexBuilder.makeInputRef(
+                  right.getRowType().getFieldList().get(i).getType(), i + fieldCount))));
+    }
+    RexNode condition = RexUtil.composeConjunction(rexBuilder, conditions);
+
+    builder.push(left)
+        .push(right)
+        .join(JoinRelType.SEMI, condition)
+        .distinct();
+
+    List<RexNode> projects = new ArrayList<>();
+    for (RexNode rexNode : builder.fields()) {
+      RelDataType originalType =
+          oriRowType.getFieldList().get(projects.size()).getType();
+      if (!originalType.equals(rexNode.getType())) {
```

Comment:
In the setop operator, it is common to encounter a mixture of Java type and SQL type.

Reply:
I am not sure what this statement means.
In our compiler we never use Java types, so they never show up in the Rel tree.
And I am not sure casts between Java types and SQL types are even legal.

Reply:
Here I refer to the logic of the existing code, if I remember correctly it should be AggJoinTranspose.

Reply:
I still don't understand what the relationship is between types and rewrite rules.

Reply:
Normally the validator inserts casts to ensure that all collections in an intersect have the same types.
(Assuming you use type coercion).

Reply:
I am also confused, but it happens, usually when operating set operators